### PR TITLE
Notification and approval of farmer edition when done by a staff member

### PR DESCRIPTION
**Pull request by daquinons**
# Description

#Motivation
Moses wants to approve the changes made by his employees because they tend to make mistakes when changing farmer data

#Finish Line
- Incoming changes by employees get stored in the DB as change requests
- Incoming changes by admins are accepted directly
- BE sends notification to FE that changes are pending (Stretch: Real-time)
- BE has endpoint to retrieve change-requests, only accessible for admins

## Checklist

Remove any items which are not applicable.

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works AND the tests pass

### CODE REVIEW

**Comment by quansenB** [comment]
Changed to Readyy-for-review now. We still miss some tests but will implement them in another PR, to accelerate deployment and FE-related work

File: models/farmer.js [hunk 1]
```diff
@@ -229,4 +229,4 @@ const farmerSchema = mongoose.Schema({
 
 const Farmer = mongoose.model('Farmer', farmerSchema);
 
-module.exports = Farmer;
+module.exports = { Farmer, farmerSchema };
```
Comment (Pav0l) [inline models/farmer.js]: Can't find this `farmerSchema` used anywhere. Am I missing something?

File: middlewares/Validations/farmer/changeRequest.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
+const Joi = require("@hapi/joi");
+const joiValidate = require("../../../helpers/joiValidate");
+
+/**
+ * Create user validation schema
```
Comment (Pav0l) [inline middlewares/Validations/farmer/changeRequest.js]: Probably typo/ctrl-c ctrl-v error. Should be `farmer` schema for change requests

File: controllers/farmer/updateFarmer.js [hunk 3]
```diff
@@ -1,9 +1,10 @@
 const { models } = require('../../models');
+const convertToDotNotationObject = require('./convertToDotNotationObject');
 const {
   createError,
   GENERIC_ERROR,
   NOT_FOUND
-} = require('../../helpers/error');
+} = require("../../helpers/error");
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/farmer/updateFarmer.js [hunk 4]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
     const { username, isAdmin } = req.user;
 
-    const farmerExist = await models.Farmer.findOne({ _id: farmerId });
-    if (!farmerExist) {
+    const farmer = await models.Farmer.findOne({ _id: farmerId });
+    if (!farmer) {
       return next(
         createError({
-          message: 'Farmer does not exist',
+          message: "Farmer does not exist",
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/farmer/updateFarmer.js [hunk 5]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
-        farmer
+        message:
+          "You are not an admin, your change was created and is ready for admin approval",
+        farmerEditRequest
       });
     }
     return next(
       createError({
-        message: 'Not authorized to update farmer details',
+        message: "Not authorized to update farmer details",
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/farmer/updateFarmer.js [hunk 6]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
         status: NOT_FOUND
       })
     );
   } catch (err) {
     return next(
       createError({
-        message: 'Could not update farmer details',
+        message: "Could not update farmer details",
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/farmer/updateFarmer.js [hunk 7]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
+        farmer_name: `${farmer.personalInfo.first_name} ${farmer.personalInfo.surname}`,
+        change_requested_by: username
+      });
 
       return res.status(201).json({
         success: true,
-        message: 'Farmer details updated successfully',
-        farmer
+        message:
+          "You are not an admin, your change was created and is ready for admin approval",
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/farmer/updateFarmer.js [hunk 8]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
+          success: true,
+          message: "Farmer details updated successfully",
+          farmer
+        });
+      }
+
+      const farmerEditRequest = await models.ChangeRequest.create({
+        requested_changes: farmerDetails,
+        farmer_id: farmerId,
+        farmer_name: `${farmer.personalInfo.first_name} ${farmer.personalInfo.surname}`,
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: I think we should have `middle_name` here as well just to be complete

File: controllers/farmer/updateFarmer.js [hunk 9]
```diff
@@ -18,40 +19,55 @@ const updateFarmer = async (req, res, next) => {
...
+        );
 
-      const farmer = await models.Farmer.findOneAndUpdate(
-        { _id: farmerId },
-        farmerDetails,
-        { new: true }
-      );
+        return res.status(201).json({
+          success: true,
+          message: "Farmer details updated successfully",
```
Comment (Pav0l) [inline controllers/farmer/updateFarmer.js]: Single quotes 🙏

File: controllers/changeRequest/getChangeRequests.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    const selectFilter = Object.keys(
+      convertToDotNotationObject(requested_changes)
+    );
+    const original_data = await models.Farmer.findOne(
+      {
+        _id: changeRequest.farmer_id
+      },
+      selectFilter
+    ).lean();
+    delete original_data._id;
```
Comment (Pav0l) [inline controllers/changeRequest/getChangeRequests.js]: Nice work with `lean()` method. TIL! 🔥

File: controllers/changeRequest/getChangeRequests.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      _id: changeRequest._id,
+      farmer_name: changeRequest.farmer_name,
+      change_requested_by: changeRequest.change_requested_by,
+      datetime: changeRequest.date
+    };
+
+    return res.status(200).json({
+      success: true,
+      message: 'ChangeRequest with this ID found',
+      ...responseObject
```
Comment (Pav0l) [inline controllers/changeRequest/getChangeRequests.js]: What is the purpose of `...` here?

File: tests/changeRequest.test.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+chai.use(chaiHttp);
+
+describe('Request change route', () => {
+  let token = '';
+  let staffToken = '';
+  let idCreatedByStaff = '';
+  let changeRequestId = '';
+
+  it('Login user responds with 200', done => {
```
Comment (Pav0l) [inline tests/changeRequest.test.js]: 'Login admin user...'

File: tests/changeRequest.test.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+      .post(`/api/v1/change-requests/${changeRequestId}/decline`)
+      .set('Authorization', token)
+      .end((err, res) => {
+        res.should.have.status(200);
+        res.body.message.should.equal('ChangeRequest declined');
+        done(err);
+      });
+  });
+
+  it('It creates a changerequest if farmer is updated by staff', done => {
```
Comment (Pav0l) [inline tests/changeRequest.test.js]: Isn't this test a copy of test at line 55?

**Pull request merged**